reframe/REFRAME_HTTP_TRANSPORT.md

```diff
@@ -0,0 +1,103 @@
+# ![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square) Reframe: HTTP Transport
+
+**Author(s)**:
+- Adin Schmahmann
+- Petar Maymounkov
+- Marcin Rataj
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+The Reframe over HTTP protocol is defining the transport and message
+serialization mechanisms for sending Reframe messages over HTTP `POST` and
+`GET`, and provides guidance for implementers around HTTP caching.
+
+# Organization of this document
+
+- [HTTP Transport Design](#http-transport-design)
+  - [HTTP Caching Considerations](#http-caching-considerations)
+    - [POST vs GET](#post-vs-get)
+    - [Avoiding sending the same response messages twice](#avoiding-sending-the-same-response-messages-twice)
+    - [Client controls for time-based caching](#client-controls-for-time-based-caching)
+    - [Rate-limiting non-cachable POST requests](#rate-limiting-non-cachable-post-requests)
+- [Implementations](#implementations)
+
+# HTTP Transport Design
+
+All messages sent in HTTP body MUST be encoded as DAG-JSON and use explicit content type `application/vnd.ipfs.rpc+dag-json; version=1`
+
+Requests MUST be sent as either:
+- `GET /reframe/{mbase64url-dag-cbor}`
+  - Cachable HTTP `GET` requests with message passed as DAG-CBOR in HTTP path segment, encoded as URL-safe [`base64url` multibase](https://docs.ipfs.io/concepts/glossary/#base64url) string
+  - Suitable for sharing links, sending smaller messages, and when a query result MUST benefit from HTTP caching (see _HTTP Caching Considerations_ below).
+- `POST /reframe`
+  - Ephemeral HTTP `POST` request with message passed as DAG-JSON in HTTP request body
+  - Suitable for bigger messages, and when HTTP caching should be skipped for the most fresh results
+
+Severs MUST support `GET` for methods marked as cachable and MUST support `POST` for all methods (both cachable and not-cachable). This allows servers to rate-limit `POST` when cachable `GET` could be used instead, and enables clients to use `POST` as a fallback in case there is a technical problem with bigger Reframe messages not fitting in a `GET` URL. See "Caching Considerations" section.

```

Comment:
```suggestion
Servers MUST support `GET` for methods marked as cachable and MUST support `POST` for all methods (both cachable and not-cachable). This allows servers to rate-limit `POST` when cachable `GET` could be used instead, and enables clients to use `POST` as a fallback in case there is a technical problem with bigger Reframe messages not fitting in a `GET` URL. See "Caching Considerations" section.

```